core/src/services/mongodb/backend.rs

```diff
@@ -0,0 +1,315 @@
+        f.debug_struct("Adapter")
+            .field("connection_string", &self.connection_string)
+            .field("database", &self.database)
+            .field("collection", &self.collection)
+            .field("key_field", &self.key_field)
+            .field("value_field", &self.value_field)
+            .finish()
+    }
+}
+
+impl Adapter {
+    async fn get_options(&self) -> Result<&ClientOptions> {
+        self.options
+            .get_or_try_init(|| async {
+                let options = ClientOptions::parse(&self.connection_string)
+                    .await
+                    .map_err(parse_mongodb_error)?;
+                Ok(Arc::new(options))
+            })
+            .await
+            .map(|v| v.as_ref())
+    }
+    async fn get_client(&self) -> Result<&mongodb::Client> {
+        self.client
+            .get_or_try_init(|| async {
+                let client_options = self.get_options().await?.clone();
+                let client =
+                    mongodb::Client::with_options(client_options).map_err(parse_mongodb_error)?;
+                Ok(Arc::new(client))
+            })
+            .await
+            .map(|v| v.as_ref())
+    }
+
+    async fn get_database(&self) -> Result<&mongodb::Database> {
+        self.database_instance
+            .get_or_try_init(|| async {
+                let client = self.get_client().await?;
+                let database = client.database(&self.database);
+                Ok(Arc::new(database))
+            })
+            .await
+            .map(|v| v.as_ref())
+    }
+
+    async fn get_collection(&self) -> Result<&mongodb::Collection<Document>> {
+        self.collection_instance
+            .get_or_try_init(|| async {
+                let database = self.get_database().await?;
+                let collection = database.collection(&self.collection);
+                Ok(Arc::new(collection))
+            })
+            .await
+            .map(|v| v.as_ref())
+    }
+}
+
+#[async_trait]
+impl kv::Adapter for Adapter {
+    fn metadata(&self) -> kv::Metadata {
+        kv::Metadata::new(
+            Scheme::Mongodb,
+            format!("{}/{}", self.database, self.collection).as_str(),
+            Capability {
+                read: true,
+                write: true,
+                ..Default::default()
+            },
+        )
+    }
+
+    async fn get(&self, path: &str) -> Result<Option<Vec<u8>>> {
+        let collection = self.get_collection().await?;
+        let filter = doc! {self.key_field.as_str():path};
+        let result = collection
+            .find_one(filter, None)
+            .await
+            .map_err(parse_mongodb_error)?;
+        match result {
+            Some(doc) => {
+                let value = doc
+                    .get(self.value_field.as_str())
+                    .ok_or(Error::new(ErrorKind::Unexpected, "value not found"))?;
+                let value = serde_json::to_vec(value).map_err(parse_json_error)?;
```

Comment:
It doesn't make sense to me to calling `serde_json` on this value. Maybe we should use `get_binary_generic` of bson?